src/recode.jl

```diff
@@ -0,0 +1,382 @@
+        end
+        drefs[i] = v
+    end
+
+    dest
+end
+
+"""
+    recode!(a::AbstractArray[, default::Any], pairs::Pair...)
+
+Convenience function for in-place recoding, equivalent to `recode!(a, a, ...)`.
+
+# Examples
+```jldoctest
+julia> x = collect(1:10);
+julia> recode!(x, 1=>100, 2:4=>0, [5; 9:10]=>-1);
+julia> x
+10-element Array{Int64,1}:
+ 100
+   0
+   0
+   0
+  -1
+   6
+   7
+   8
+  -1
+  -1
+ ```
+"""
+recode!(a::AbstractArray, default::Any, pairs::Pair...) = recode!(a, a, default, pairs...)
+recode!(a::AbstractArray, pairs::Pair...) = recode!(a, a, nothing, pairs...)
+
+promote_valuetype{K, V}(x::Pair{K, V}) = V
+promote_valuetype{K, V}(x::Pair{K, V}, y::Pair...) = promote_type(V, promote_valuetype(y...))
+
+"""
+    recode(a::AbstractArray[, default::Any], pairs::Pair...)
+
+Return a new categorical array with elements from `a`, replacing elements matching a key
+of `pairs` with the corresponding value. The type of the array is chosen so that it can
+hold all recoded elements (but not necessarily original elements from `a`).
+
+For each `Pair` in `pairs`, if the element is equal to (according to `isequal`) or `in` the key
+(first item of the pair), then the corresponding value (second item) is copied to `src`.
+If the element matches no key and `default` is not provided or `nothing`, it is copied as-is;
+if `default` is specified, it is used in place of the original element.
+If an element matches more than one key, the first match is used.
+
+# Examples
+```jldoctest
+julia> recode(1:10, 1=>100, 2:4=>0, [5; 9:10]=>-1)
+10-element Array{Int64,1}:
+ 100
+ 0  
+ 0  
+ 0  
+ -1 
+ 6  
+ 7  
+ 8  
+ -1 
+ -1 
+ ```
+
+     recode(a::AbstractArray{<:Nullable}[, default::Any], pairs::Pair...)
+
+For a nullable array `a`, automatic lifting is applied: values are unwrapped before
+comparing them to the keys of `pairs` using `isequal`. Null values are never replaced
+with `default`: use `Nullable()` in a pair to recode them.
+
+Return a `NullableArray` unless `default` is provided and not nullable, in which case
+an `Array` is returned.
+
+# Examples
+```jldoctest
+julia> using NullableArrays
+
+julia> recode(NullableArray(1:10), 1=>100, 2:4=>0, [5; 9:10]=>-1, 6=>Nullable())
+10-element NullableArrays.NullableArray{Int64,1}:
+ 100  
+ 0    
+ 0    
+ 0    
+ -1   
+ #NULL
+ 7    
+ 8    
+ -1   
+ -1   
+ ```
+"""
+function recode end
+
+recode(a::AbstractArray, pairs::Pair...) = recode(a, nothing, pairs...)
+
+function recode(a::AbstractArray, default::Any, pairs::Pair...)
+    V = promote_valuetype(pairs...)
+    # T cannot take into account eltype(src), since we can't know
+    # whether it matters at compile time (all levels recoded or not)
+    # and using a wider type than necessary would be annoying
+    T = default === nothing ? V : promote_type(typeof(default), V)
+    # TODO: result should not be a nullable array when one of the pairs' LHS is null
+    if T <: Nullable || eltype(a) <: Nullable
```

Comment:
Yeah, that's a bit problematic. In theory we should use `typeintersect`, but in practice it doesn't sound very useful to return a `NullableArray` when `T == Any`.

The TODO refers to the fact that with `Union{T, Null}`, we will be able to know whether one of the pairs' LHS is `null` at compile time just from its type, which isn't possible with `Nullable`. This will allow returning a non-nullable array when we know in advance all missing values will be replaced with a non-null value.